### PR TITLE
Update API usage to OpenSSL 3 (RSA)

### PR DESCRIPTION
**Pull request by iamnotacake**
Second part regarding changes that replace deprecated OpenSSL API usage with new alternatives from v3. This time related to RSA keys serialization/deserialization, and few other related places.

## Checklist

- [x] Change is covered by automated tests
- [ ] Benchmark results are attached (if applicable)
- [x] The [coding guidelines] are followed
- [x] ~Public API has proper documentation~
- [x] ~Example projects and code samples are up-to-date (in case of API changes)~
- [ ] Changelog is updated (in case of notable or breaking changes)

[coding guidelines]: https://github.com/cossacklabs/themis/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: src/soter/openssl/soter_rsa_key_3.c
Comment (Lagovas) [inline src/soter/openssl/soter_rsa_key_3.c:349]: great comments

File: src/soter/openssl/soter_rsa_key_3.c
Comment (Lagovas) [inline src/soter/openssl/soter_rsa_key_3.c:493]: yes, if we use it in several places. as I understand, previously it was used only in one place?

File: src/soter/openssl/soter_rsa_key_3.c
Comment (Lagovas) [inline src/soter/openssl/soter_rsa_key_3.c:675]: ?

File: src/soter/openssl/soter_rsa_key_3.c
Comment (Lagovas) [inline src/soter/openssl/soter_rsa_key_3.c:760]: start of the rsa configuration symphony)

File: src/soter/openssl/soter_asym_cipher.c
Comment (iamnotacake) [reply src/soter/openssl/soter_asym_cipher.c]: Done

File: src/soter/openssl/soter_asym_cipher.c
Comment (iamnotacake) [reply src/soter/openssl/soter_asym_cipher.c]: Done. Although `man` says `EVP_PKEY_private_check` was added in 3.0, so cannot use for 1.X

File: src/soter/openssl/soter_rsa_key_3.c
Comment (iamnotacake) [reply src/soter/openssl/soter_rsa_key_3.c:99]: Well, `bignum_to_bytes` and `bn_encode` are not exactly the same

File: src/soter/openssl/soter_rsa_key_3.c
Comment (iamnotacake) [reply src/soter/openssl/soter_rsa_key_3.c]: Done

File: src/soter/openssl/soter_rsa_key_3.c
Comment (iamnotacake) [reply src/soter/openssl/soter_rsa_key_3.c:299]: This TODO was in original `soter_rsa_key.c` as well. Is it worth touching in _this_ PR?

File: src/soter/openssl/soter_rsa_key_3.c
Comment (iamnotacake) [reply src/soter/openssl/soter_rsa_key_3.c]: Small mistake, not caught by the compiler, removed

File: src/soter/openssl/soter_rsa_key_3.c
Comment (iamnotacake) [reply src/soter/openssl/soter_rsa_key_3.c:565]: This makes me think whether `EVP_PKEY_todata()` could be used in RSA private key serialization. Since it got a lot of params being extracted, and if `_todata()` could do it in less number of function calls, it may be more convenient than current approach. Public key on the other hand only got two properties. And EC keys only have one property being read

File: src/soter/openssl/soter_rsa_key_3.c
Comment (iamnotacake) [reply src/soter/openssl/soter_rsa_key_3.c]: UPD: Removed both of these functions and put `BN_bn2binpad()` instead. Looks cleaner, works like before

File: src/soter/openssl/soter_rsa_key_3.c
Comment (ilammy) [inline src/soter/openssl/soter_rsa_key_3.c:409]: `free_builder` label is never used.

Since those functions follow free() convention, you can use a single label `out` or `err`.

**Comment by ilammy** [comment]
Tests pass ☀️

LGTM, provided a decision on OpenSSL 1.0.2 support is made (followed by some reverts, if we still support it).

File: src/soter/openssl/soter_ec_key.c
Comment (Lagovas) [reply src/soter/openssl/soter_ec_key.c:160]: looks like we should use the previously removed `bignum_to_bytes` function which will encapsulate these ifdefs with the comment of removing the old version of implementation after dropping openssl 1.0.2 support

File: src/soter/openssl/soter_rsa_key_3.c
Comment (Lagovas) [reply src/soter/openssl/soter_rsa_key_3.c:52]: if it wasn't done for openssl1.x.x implementation, so we should do the same for old code too

File: .github/workflows/test-core.yaml
Comment (iamnotacake) [inline .github/workflows/test-core.yaml:405]: Missed this line, will remove in next commit. BTW, why [MSYS2 environment](https://github.com/cossacklabs/themis/actions/runs/4326599795/jobs/7554211265) job fails like there is still "no support for OpenSSL 3" in header file?

File: src/soter/openssl/soter_asym_cipher.c
Comment (iamnotacake) [reply src/soter/openssl/soter_asym_cipher.c]: Done, thanks. Also created function `check_rsa_private_key`. Used only once, but has three variants for different OpenSSL versions, moved outside to make caller func look simpler. One small drawback of changes in `soter_asym_cipher_decrypt()` is that `EVP_PKEY_get0()` has to be called twice, in `get_rsa_mod_size()` and in `check_rsa_private_key()` for OpenSSL 1.X.

File: src/soter/openssl/soter_rsa_key_3.c
Comment (iamnotacake) [reply src/soter/openssl/soter_rsa_key_3.c]: Done. Missed that when moved things around to reuse single BIGINT

File: src/soter/openssl/soter_rsa_key_3.c
Comment (iamnotacake) [reply src/soter/openssl/soter_rsa_key_3.c]: Done

File: src/soter/openssl/soter_rsa_key_3.c
Comment (ilammy) [inline src/soter/openssl/soter_rsa_key_3.c:169]: Forgotten comment.

File: src/soter/openssl/soter_rsa_key_3.c
Comment (ilammy) [inline src/soter/openssl/soter_rsa_key_3.c:401]: Another now-unnecessary comment.

File: src/soter/openssl/soter_rsa_key_utils.h
Comment (ilammy) [inline src/soter/openssl/soter_rsa_key_utils.h:105]: Oh, there is no common header file to avoid duplication 😞 Well, not a big deal. Rule of three.

**Review by ilammy — APPROVED** [review]
Good job on getting all those warnings resolved! 💪

File: .github/workflows/test-core.yaml
Comment (ilammy) [reply .github/workflows/test-core.yaml:405]: > why [MSYS2 environment](https://github.com/cossacklabs/themis/actions/runs/4326599795/jobs/7554211265) job fails like there is still "no support for OpenSSL 3" in header file?

Hm... Looks like the package uses the pristine 0.14.0 tarball, not the source from the current tree.

That is, when you build a package, it's built locally but still references the tarball from release artifacts, not packages the source from whatever the current branch contains. And it's been doing that for a while. Says a lot about how well-maintained MSYS2 builds are.

Uh, well... I think you should leave `WITH_EXPERIMENTAL_OPENSSL_3_SUPPORT: yes` here for the time being. Then *someone* 🙄 should apply whatever hacks necessary for this job to package the tree being tested.

Comment (iamnotacake) [reply .github/workflows/test-core.yaml:405]: Kinda weird seeing Pacman being used on Windows. Anyway, added this line back and another comment instead, since "stop using deprecated API" is not relevant anymore

**Comment by iamnotacake** [comment]
BTW, I managed to build (and even debug) libsoter with OpenSSL 1.0.2g. Using CLion (CMake project) and Docker (Ubuntu 16.04 with compiler and openssl being added). Even though current `CMakeLists.txt` requires CMake 3.8, it worked with CMake 3.5.1 that was available on old Ubuntu. Had to lower the version and add `include_directories(include)` to make it work though.

File: src/soter/openssl/soter_rsa_key_3.c
Comment (Lagovas) [inline src/soter/openssl/soter_rsa_key_3.c:85]: should we validate `key` that it is not null before accessing? above we checked only key_length

File: src/soter/openssl/soter_rsa_key_3.c
Comment (Lagovas) [inline src/soter/openssl/soter_rsa_key_3.c:128]: same here. can you review all functions and add checks for input arguments?

File: src/soter/openssl/soter_rsa_key_utils.h
Comment (Lagovas) [inline src/soter/openssl/soter_rsa_key_utils.h:124]: same here, soter wipe

File: src/soter/openssl/soter_rsa_key_utils.h
Comment (Lagovas) [reply src/soter/openssl/soter_rsa_key_utils.h:105]: hm, I remember that we always followed rule of two ) if it repeats second time then it is a signal to refactor it and move to common file
so, can we do not duplicate it and place in the soter_bignum_utils.h? or in some common file? because if we will need to update something, we should keep in mind that we have same copy in the another file...
now we should update `memset` -> `soter_wipe` in both places...

File: src/soter/openssl/soter_ec_key_utils.h
Comment (iamnotacake) [reply src/soter/openssl/soter_ec_key_utils.h:120]: No, this `memset` is just to add zero bytes if padding is needed, this is not "erase on fail", same in `soter_rsa_key_utils.h`

File: src/soter/openssl/soter_rsa_key_3.c
Comment (iamnotacake) [reply src/soter/openssl/soter_rsa_key_3.c:85]: But we have a check above, lines 74..78
```
    if ((!key) || (output_length > *key_length)) {
        *key_length = output_length;
        res = SOTER_BUFFER_TOO_SMALL;
        goto err;
    }
```

File: src/soter/openssl/soter_rsa_key_3.c
Comment (iamnotacake) [reply src/soter/openssl/soter_rsa_key_3.c:128]: This one is kinda weird. Indeed, we assign `curr_bn = key` before checking key. But we access `curr_bn` only after checking `key`, so there is no bug here. To make it cleaner, I'll move assignment to a place between check and first usage

**Review by Lagovas — APPROVED** [review]
lgtm but lets make tests green

**Comment by Lagovas** [comment]
due to tests fails because of one dependency increased required version of cargo and this PR merges into the temporary branch before merging to master, we can merge it to it, unblock final reviewing and fixing tests in separate PR #995

**Pull request merged**
